### PR TITLE
Final status

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
this is especially useful for automated log analysis (e.g. borgweb).

if there is a definite rc 0 or rc 1 in the log, the tool can easily find out whether backup succeeded or not and indicate green or red status.

### CODE REVIEW

File: borg/archiver.py [hunk 1]
```diff
@@ -1042,22 +1043,33 @@ def main():  # pragma: no cover
...
+        logger.info(exit_msg % ('success', exit_code))
+    elif exit_code == EXIT_WARNING:
+        logger.warning(exit_msg % ('warning', exit_code))
+    elif exit_code == EXIT_ERROR:
+        logger.error(exit_msg % ('error', exit_code))
+    else:
+        logger.error(exit_msg % ('abnormal', exit_code))
     sys.exit(exit_code)
 
+
```
Comment (anarcat) [inline borg/archiver.py]: is that pep8?

File: borg/archiver.py [hunk 2]
```diff
@@ -1042,22 +1043,33 @@ def main():  # pragma: no cover
...
+        logger.error(msg)
+    exit_msg = 'terminating with %s status, rc %d'
+    if exit_code == EXIT_SUCCESS:
+        logger.info(exit_msg % ('success', exit_code))
+    elif exit_code == EXIT_WARNING:
+        logger.warning(exit_msg % ('warning', exit_code))
+    elif exit_code == EXIT_ERROR:
+        logger.error(exit_msg % ('error', exit_code))
+    else:
+        logger.error(exit_msg % ('abnormal', exit_code))
```
Comment (anarcat) [inline borg/archiver.py]: >  codecov/patch — 67.00% of diff hit (target 88.00%)

tests should be expanded to cover for these scenarios...

File: borg/helpers.py [hunk 3]
```diff
@@ -487,10 +501,6 @@ def format_archive(archive):
     return '%-36s %s' % (archive.name, to_localtime(archive.ts).strftime('%c'))
 
 
-class IntegrityError(Error):
```
Comment (anarcat) [inline borg/helpers.py]: why was this moved around?

File: borg/helpers.py [hunk 4]
```diff
@@ -44,13 +44,27 @@ def have_cython():
     import msgpack
 
 
+# return codes returned by borg command
+# when borg is killed by signal N, rc = 128 + N
+EXIT_SUCCESS = 0  # everything done, no problems
+EXIT_WARNING = 1  # reached normal end of operation, but there were issues
+EXIT_ERROR = 2  # terminated abruptly, did not reach end of operation
```
Comment (anarcat) [inline borg/helpers.py]: wouldn't it be worthwhile to move all the error handling code to a error.py module?

File: borg/helpers.py (hunk 3 above)
Comment (ThomasWaldmann) [inline borg/helpers.py]: so both classes deriving from Error are together. before it was just at some random position.

File: borg/helpers.py (hunk 4 above)
Comment (ThomasWaldmann) [inline borg/helpers.py]: no, currently "Error" exceptions are defined where needed.

what I am thinking about implementing at some time is a "constants" module that just has all these constants.

File: borg/archiver.py (hunk 1 above)
Comment (ThomasWaldmann) [inline borg/archiver.py]: yes, 2 lines between toplevel stuff.

File: borg/archiver.py (hunk 2 above)
Comment (ThomasWaldmann) [inline borg/archiver.py]: this is in main and it sys.exits there, maybe difficult to test.
also maybe not worth it.

Comment (anarcat) [inline borg/archiver.py]: but we do fork and test borg commands right now... no? we could check return values...

File: borg/helpers.py (hunk 3 above)
Comment (anarcat) [inline borg/helpers.py]: okay

File: borg/helpers.py (hunk 4 above)
Comment (anarcat) [inline borg/helpers.py]: but those are all defined in `helpers.py`... are they only for stuff in `helpers.py`?

i agree that constants should be defined in the relevant module however... so maybe disregard this if those are `helpers`-specific errors.

Comment (ThomasWaldmann) [inline borg/helpers.py]: these constants are needed by misc modules and the general policy is to have random small stuff needed from everywhere in helpers as far as i can tell.

Comment (anarcat) [inline borg/helpers.py]: okay, makes sense for now then.

File: borg/archiver.py (hunk 2 above)
Comment (ThomasWaldmann) [inline borg/archiver.py]: the archiver tests all run in misc. ways:
- local repo vs. remote repo
- not forked (python code) vs. forked (binary)

so, if we use the binary or maybe better the installed borg script, we could test 3 return codes.

Comment (anarcat) [inline borg/archiver.py]: ... i guess we should check the return value of the fixture calls (they return the exit code) or exec_cmd() when we call it directly...
